content/blog/ditch-your-version-manager/index.md

```diff
@@ -0,0 +1,336 @@
+
+```bash
+$ echo "use nix" > .envrc
+direnv: error /Users/arkham/Desktop/nix_hello_world/.envrc
+is blocked. Run `direnv allow` to approve its content
+
+$ direnv allow
+<loads of nix output>
+
+$ hello
+Hello, world!
+```
+
+🎉
+
+What did we do?
+
+- we imported the whole set of packages provided by Nix and called it `pkgs`
+- we used the `mkShell` built-in function to create a new environment where
+  we can use Nix programs
+- we specify we want the GNU `hello` program
+
+What if we wanted ruby? Couldn't be simpler:
+
+```nix
+let
+  pkgs = import <nixpkgs> { };
+in
+pkgs.mkShell {
+  buildInputs = [
+    pkgs.hello
+    pkgs.ruby
+  ];
+}
+```
+
+By saving this file and returning to your shell (you might need to press
+`Enter` to let direnv reload), Nix will pick up the change and install the
+new dependency.
+
+```bash
+$ ruby --version
+ruby 2.7.4p191 (2021-07-07) [x86_64-darwin17]
+
+$ which ruby
+/nix/store/r6siyyqmnidkn2y8y5hl7payykb51kb0-ruby-2.7.4/bin/ruby
+```
+
+What's with that super long path? Don't worry, it's just how Nix stores its
+data (*waves hands.. iMmUTaBIlitY*).
+
+What if we needed node?
+
+```nix
+let
+  pkgs = import <nixpkgs> { };
+in
+pkgs.mkShell {
+  buildInputs = [
+    pkgs.hello
+    pkgs.ruby
+    pkgs.nodejs
+  ];
+}
+```
+
+```bash
+$ node --version
+v14.17.2
+
+$ which node
+/nix/store/zj1sqykb3s5a0fakm94qwz7fg6g2zxpm-nodejs-14.17.2/bin/node
+```
+
+You see the pattern here.
+
+How about a different version of Ruby or Node? Let's say that our project
+depends on Ruby 2.6 and Node 10. We can [go and
+search](https://search.nixos.org/packages) for those specific versions,
+then change our `default.nix` accordingly:
+
+```nix
+let
+  pkgs = import <nixpkgs> { };
+in
+pkgs.mkShell {
+  buildInputs = [
+    pkgs.hello
+    pkgs.ruby_2_6
+    pkgs.nodejs-10_x
+  ];
+}
+```
+
+```
+$ ruby --version
+ruby 2.6.8p205 (2021-07-07) [x86_64-darwin17]
+
+$ node --version
+v10.24.1
+```
+
+🎉
+
+## Some bad news
+
+Unfortunately, if you tried to follow this article step by step, you'll
+have noticed that the versions of `ruby` and `node` you installed are
+probably slightly different from the ones above.
+
+Wasn't Nix supposed to give us _reproducible_ builds?
+
+Yes, but...
+
+All the software that we installed depends on the specific version of the
+**nixpkgs channel** that we installed on our system. We can check which
+version that is by running:
+
+```bash
+$ nix repl
+Welcome to Nix version 2.3.10. Type :? for help.
+
+nix-repl> (import <nixpkgs> {}).lib.version
+"21.05pre292875.60563458051"
+```
+
+We can then grab the commit part (the third segment) and view it on
+[GitHub](https://github.com/NixOS/nixpkgs/commit/60563458051).
+
+If only we were pointing to the same commit, then we would have exactly the
+same dependencies. But it would be foolish to require every project to
+depend on the same version of nixpkgs. We need a way to specify this on a
+per-project basis, right?
```

Comment:
Not a big deal but I wonder if this would read nicer if you had:

```
We need a way to specify this on a per-project basis, right? We can use a tool called `niv` to help us with this.

## Niv

[niv](https://github.com/nmattia/niv) lets us pin our nixpkgs configuration to a specific version.
```